strangelove-ventures/cosmos-operator
# feat(CosmosFullNode): Add private peer information to ConfigMap

Closes https://github.com/strangelove-ventures/cosmos-operator/issues/42

This completes the ability for each pod to peer with each other privately.

## Known Limitations

* If the user adds their own unconditional or private peer ids by using toml overrides, it will clobber what the operator creates. 
* Scaling replicas up or down will always cause a rollout because the config changes to account for increase/decreased peers. I think this is an acceptable tradeoff. 

## QA

Confirmed working via logs. 

`c8ece62213715696e9561d7e03648306ecb67749` is the private peer of the other pod. 

In these logs, I purposefully rebooted the other pod. The logs indicate it was disconnected and later reconnected. 

```
default cosmoshub-dev-0 node 3:18PM INF Adding persistent peers addrs=["c8ece62213715696e9561d7e03648306ecb67749@cosmoshub-dev-p2p-1.default.svc.cluster.local:26656"] module=p2p
default cosmoshub-dev-0 node 3:18PM INF Adding unconditional peer ids ids=["c8ece62213715696e9561d7e03648306ecb67749"] module=p2p
default cosmoshub-dev-0 node 3:18PM INF service start impl="Peer{MConn{10.124.11.249:26656} c8ece62213715696e9561d7e03648306ecb67749 out}" module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656}
default cosmoshub-dev-0 node 3:18PM INF service start impl=MConn{10.124.11.249:26656} module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656}
default cosmoshub-dev-0 node 3:19PM INF Connection is closed @ recvRoutine (likely by the other side) conn={"Logger":{}} module=p2p peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656}
default cosmoshub-dev-0 node 3:19PM INF service stop impl={"Logger":{}} module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656}
default cosmoshub-dev-0 node 3:19PM INF service stop impl={"Data":{},"Logger":{}} module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656}
default cosmoshub-dev-0 node 3:19PM INF Reconnecting to peer addr={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656} module=p2p
default cosmoshub-dev-0 node 3:19PM INF Error reconnecting to peer. Trying again addr={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656} err="dial tcp 10.124.11.249:26656: connect: connection refused" module=p2p tries=0
default cosmoshub-dev-0 node 3:19PM INF Error reconnecting to peer. Trying again addr={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656} err="dial tcp 10.124.11.249:26656: connect: connection refused" module=p2p tries=1
default cosmoshub-dev-0 node 3:19PM INF Error reconnecting to peer. Trying again addr={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656} err="dial tcp 10.124.11.249:26656: connect: connection refused" module=p2p tries=2
default cosmoshub-dev-0 node 3:19PM INF Error reconnecting to peer. Trying again addr={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656} err="dial tcp 10.124.11.249:26656: connect: connection refused" module=p2p tries=3
default cosmoshub-dev-0 node 3:19PM INF Error reconnecting to peer. Trying again addr={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656} err="dial tcp 10.124.11.249:26656: connect: connection refused" module=p2p tries=4
default cosmoshub-dev-0 node 3:19PM INF Error reconnecting to peer. Trying again addr={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.124.11.249","port":26656} err="dial tcp 10.124.11.249:26656: connect: connection refused" module=p2p tries=5
default cosmoshub-dev-0 node 3:19PM INF service start impl="Peer{MConn{10.120.2.107:55136} c8ece62213715696e9561d7e03648306ecb67749 in}" module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.120.2.107","port":55136}
default cosmoshub-dev-0 node 3:19PM INF service start impl=MConn{10.120.2.107:55136} module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.120.2.107","port":55136}
default cosmoshub-dev-0 node 3:33PM ERR Failed to write PacketMsg err="write tcp 10.120.0.18:26656->10.120.2.107:55136: write: connection reset by peer" module=p2p peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.120.2.107","port":55136}
default cosmoshub-dev-0 node 3:33PM INF service stop impl={"Logger":{}} module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.120.2.107","port":55136}
default cosmoshub-dev-0 node 3:33PM INF service stop impl={"Data":{},"Logger":{}} module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.120.2.107","port":55136}
default cosmoshub-dev-0 node 3:33PM INF service start impl="Peer{MConn{10.120.2.108:39362} c8ece62213715696e9561d7e03648306ecb67749 in}" module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.120.2.108","port":39362}
default cosmoshub-dev-0 node 3:33PM INF service start impl=MConn{10.120.2.108:39362} module=p2p msg={} peer={"id":"c8ece62213715696e9561d7e03648306ecb67749","ip":"10.120.2.108","port":39362}
```

It also appears in `net_info`. 

```
❮ curl http://localhost:26657/net_info\? | jq -r '.result.peers[].node_info' | rg -C 10 c8ece62213715696e9561d7e03648306ecb67749
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 68694    0 68694    0     0   331k      0 --:--:-- --:--:-- --:--:--  335k
    "tx_index": "off",
    "rpc_address": "tcp://127.0.0.1:26657"
  }
}
{
  "protocol_version": {
    "p2p": "8",
    "block": "11",
    "app": "0"
  },
  "id": "c8ece62213715696e9561d7e03648306ecb67749",
  "listen_addr": "tcp://0.0.0.0:26656",
  "network": "cosmoshub-4",
  "version": "v0.34.26",
  "channels": "40202122233038606100",
  "moniker": "cosmoshub-dev-1",
  "other": {
    "tx_index": "on",
    "rpc_address": "tcp://0.0.0.0:26657"
  }
}
```
